src/test/ui/lint/opaque-ty-ffi-unsafe.stderr

```diff
@@ -1,4 +1,4 @@
-error: `extern` block uses type `impl Fn<()>`, which is not FFI-safe
+error: `extern` block uses type `impl Fn<()> + FnOnce<(), Output = ()>`, which is not FFI-safe
```

Comment:
This is what I was mentioning. I think that we can look at the traits to see that one implies the other to show only one at a time, without special casing closures.

Reply:
Yeah so this is printing because we have two predicates here `Fn<()>` and `<Self as FnOnce<()>>::Output == ()`. The first one makes us print `Fn<()>` and the second makes us print `FnOnce<(), Output = ()>`.

I actually would be interested in special casing just `Fn`/`FnMut`/`FnOnce` and fixing it up further so it prints `impl Fn() -> ()` instead.

Reply:
You can keep track of the traits for the predicates you've seen already and see if the new predicate's trait is part of the existing ones' super-traits. Special casing sounds reasonable-ish, but I would prefer it if we could just make it work for the general case :)

Regardless, dealing with the `<_ as async {}>::Return` types in a nicer way (even if it is showing them as `_`) has priority over this.

Reply:
> You can keep track of the traits for the predicates you've seen already

Is there any ordering enforced for the list of predicates that we're iterating over here, though? 

And I'm pretty sure this only prints supertraits if you're bounding an associated type on that supertrait, so in practice we really only see this with `Fn*` family of traits I think. (Because `impl Fn() -> Foo` desugars to `impl Fn<()> + FnOnce<(), Output=Foo>`, or similar, right?)

Reply:
Closures is gonna be the most common thing people will see, but a crate can write their own API structure that does the same thing. If closures are special cased, any user of those (potential, as none come to mind) crates could get confused.